src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java

```diff
@@ -611,4 +625,59 @@ private PocketChangeEvent createPocketChangeEvent(Pocket pocket, BigDecimal amou
     private BigDecimal getAmountAfter(boolean isSubtractAmount, BigDecimal balanceAmount, BigDecimal deltaAmount) {
         return isSubtractAmount ? balanceAmount.subtract(deltaAmount) : balanceAmount.add(deltaAmount);
     }
+
+    @Transactional
+    public BalanceDTO updateStatus(Long id, String status, Map<String, Object> context) {
+        log.info("Start update balance status: id: {}, status: {}, context: {}", id, status, context);
+        Balance balance = getBalanceForUpdate(id);
+        return self.updateStatus(balance, status, context);
+    }
+
+    @Transactional
+    @LogicExtensionPoint(value = "ChangeStatus", resolver = BalanceTypeKeyResolver.class)
+    public BalanceDTO updateStatus(Balance balance, String status, Map<String, Object> context) {
+        assertStatusTransition(status, balance);
+        BalanceChangeEvent changeEvent = createChangeStatusChangeEvent(balance, status);
+        log.info("Update balance status changeEvent: {}", changeEvent);
+        balanceChangeEventRepository.save(changeEvent);
+
+        balance.setStatus(status);
+        log.info("Update balance: {}", balance);
+        balance = balanceRepository.save(balance);
+
+        return balanceMapper.toDto(balance);
+    }
+
+    private void assertStatusTransition(String newStatus, Balance balance) {
+        String currentStatus = balance.getStatus();
+        String balanceTypeKey = balance.getTypeKey();
+
+        if (currentStatus == null) {
+            assertNullStatusTransition(newStatus, balanceTypeKey);
+            return;
+        }
+
+        Optional<List<NextSpec>> nextSpecs = balanceSpecService.nextSpecs(balanceTypeKey, currentStatus);
+        if (nextSpecs.isEmpty() || nextSpecs.get().stream().map(NextSpec::getStatusKey).noneMatch(newStatus::equals)) {
+            throw new StatusTransitionException(newStatus, currentStatus, balanceTypeKey);
+        }
+    }
+
+    private void assertNullStatusTransition(String newStatus, String balanceTypeKey) {
+        List<String> balanceStatusKeys = balanceSpecService.balanceStatusKeys(balanceTypeKey);
+        if(!balanceStatusKeys.contains(newStatus)){
+            throw new StatusTransitionException(newStatus, null, balanceTypeKey);
+        }
+    }
+
+    private BalanceChangeEvent createChangeStatusChangeEvent(Balance balance, String status) {
+        String operationUuid = UUID.randomUUID().toString();
+        Instant operationDate = now(clock);
+        Map<String, String> metadata = new HashMap<>();
+        metadata.put("change_status_from", balance.getStatus());
```

Comment:
let's have some `MetadataConstants` and define these `change_status..` there

Reply:
fixed